system/lib/wasmfs/file.h

```diff
@@ -8,37 +8,123 @@
 
 #pragma once
 
+#include <assert.h>
 #include <emscripten/html5.h>
 #include <mutex>
+#include <unordered_map>
 #include <vector>
 #include <wasi/api.h>
 
 namespace wasmfs {
 
-class File {
-  // TODO: Add other File properties later.
+class File : public std::enable_shared_from_this<File> {
+
+public:
+  enum FileKind { DataFileKind = 0, DirectoryKind, SymlinkKind };
+
+  // curr->is<Directory>()
+  template <class T> bool is() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::expectedKind);
+  }
+
+  template <class T> T* dynCast() {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::expectedKind) ? (T*)this : nullptr;
+  }
 
+  template <class T> T* cast() {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    assert(int(kind) == int(T::expectedKind));
+    return (T*)this;
+  }
+
+  class Handle {
+
+    std::unique_lock<std::mutex> lock;
+
+  protected:
+    std::shared_ptr<File> file;
+
+  public:
+    Handle(std::shared_ptr<File> file) : file(file), lock(file->mutex) {}
+  };
+
+  Handle get() { return Handle(shared_from_this()); }
+
+protected:
+  File(FileKind kind) : kind(kind) {}
   // A mutex is needed for multiple accesses to the same file.
   std::mutex mutex;
 
+private:
+  // TODO: Add other File properties later.
+
+  FileKind kind;
+};
+
+class DataFile : public File {
+
   virtual __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) = 0;
   virtual __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) = 0;
 
 public:
-  virtual ~File() = default;
-  class Handle {
-    File& file;
-    std::unique_lock<std::mutex> lock;
+  static constexpr FileKind expectedKind = File::DataFileKind;
+  DataFile() : File(File::DataFileKind) {}
+  virtual ~DataFile() = default;
+  class Handle : public File::Handle {
+
+    DataFile& getFile() { return *(file.get()->cast<DataFile>()); }
 
   public:
-    Handle(File& file) : file(file), lock(file.mutex) {}
+    Handle(std::shared_ptr<File> dataFile) : File::Handle(dataFile) {}
     Handle(Handle&&) = default;
 
-    __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) { return file.read(buf, len); }
-    __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) { return file.write(buf, len); }
+    __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) { return getFile().read(buf, len); }
+    __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) {
+      return getFile().write(buf, len);
+    }
+  };
+
+  Handle get() { return Handle(shared_from_this()); }
+};
+
+class Directory : public File {
+protected:
+  std::unordered_map<std::string, std::shared_ptr<File>> entries;
```

Comment:
I think this will need to be a `std::vector`, or something else ordered, because the order of iteration matters. But, that can wait for a future PR that handles getdents.